File: examples/csharp/Angzarr.Client/Descriptor.cs
namespace Angzarr.Client;

/// <summary>
/// Component type constants for descriptors.
/// </summary>
public static class ComponentTypes
{
    public const string Aggregate = "aggregate";
    public const string Saga = "saga";
    public const string ProcessManager = "process_manager";
    public const string Projector = "projector";
    public const string Upcaster = "upcaster";
}

/// <summary>
/// Describes what a component subscribes to or sends to.
/// </summary>
public record TargetDesc(string Domain, List<string> Types);

/// <summary>
/// Describes a component for topology discovery.
/// </summary>
public record Descriptor(string Name, string ComponentType, List<TargetDesc> Inputs);

File: client/java/client/src/main/java/dev/angzarr/client/Descriptor.java
package dev.angzarr.client;

import java.util.List;

/**
 * Describes what a component subscribes to or sends to.
 */
class TargetDesc {
    private final String domain;
    private final List<String> types;

    public TargetDesc(String domain, List<String> types) {
        this.domain = domain;
        this.types = types;
    }

    public String getDomain() {
        return domain;
    }

    public List<String> getTypes() {
        return types;
    }
}

/**
 * Describes a component for topology discovery.
 */
public class Descriptor {
    private final String name;
    private final String componentType;
    private final List<TargetDesc> inputs;

    public Descriptor(String name, String componentType, List<TargetDesc> inputs) {
        this.name = name;
        this.componentType = componentType;
        this.inputs = inputs;
    }

    public String getName() {
        return name;
    }

    public String getComponentType() {
        return componentType;
    }

    public List<TargetDesc> getInputs() {
        return inputs;
    }
}
